### ISSUE TITLE
Add analytical decomposition for synthesis of single qubit unitaries generated by Clifford and T gates.

### ISSUE THREAD

**Issue by tanujkhattar** [quantumlib/Cirq#5815]
**Is your feature request related to a use case or problem? Please describe.**
One of the promising gatesets for fault tolerant regime is the Clifford + T gateset, where executing T gates is significantly slower than executing Cliffords and therefore it's often desired to minimize the T-count of the compiled circuits.  Cirq currently does not have any analytical decompositions for synthesizing arbitrary 1/2q unitaries in terms of Clifford + T gateset.

**Describe the solution you'd like**
[Fast and efficient exact synthesis of single qubit unitaries generated by Clifford and T gates](https://arxiv.org/abs/1206.5236) provides an efficient algorithm for synthesis of arbitrary single qubit rotations in terms of Clifford + T gates. The feature request is to implement the described algorithm as an [analytical decomposition in Cirq](https://github.com/quantumlib/Cirq/tree/master/cirq-core/cirq/transformers/analytical_decompositions) 

**What is the urgency from your perspective for this issue? Is it blocking important work?**
P2 - we should do it in the next couple of quarters 

cc @augustehirth This would need documentation once we get this done.

**Comment by tanujkhattar** [comment]
A related recent paper for decomposing arbitrary n-qubit unitaries, which is probably out of scope for this feature request but can be done as a follow-up if we have enthusiastic contributors. 

https://link.springer.com/article/10.1007/s11128-020-02816-0

**Comment by aksaw** [comment]
I'd be very interested in giving this a shot.

**Comment by NoureldinYosri** [comment]
@aksaw Is there progress on this? I will need this soon.

**Comment by aksaw** [comment]
Yes, I had started working on this already. I can try to finish it up in the next couple of weeks.

**Comment by tanujkhattar** [comment]
@aksaw Are you still working on this?

**Comment by mhucka** [comment]
@aksaw Are you still interested in working on this?

**Comment by vivek-kumar9696** [comment]
Hi, I would like to take this up. Any immediate timelines that I need to be aware of ?

**Comment by vivek-kumar9696** [comment]
Hi, 
I would like to work on this. Before starting I wanted to confirm that we need to implement Algorithm 1 of [https://arxiv.org/pdf/1206.5236](url) right?

**Comment by NoureldinYosri** [comment]
@vivek-kumar9696 hi, sorry for the late reply ... yes that is the goal

**Comment by vivek-kumar9696** [comment]
> [@vivek-kumar9696](https://github.com/vivek-kumar9696) hi, sorry for the late reply ... yes that is the goal

Hi @NoureldinYosri,
Thanks for the reply. I would like to take on this problem, but It will take me sometime to get started with this. Is someone working on this currently?

**Comment by NoureldinYosri** [comment]
@vivek-kumar9696 thanks for offering to help. I assigned the task to you

**Comment by codrut3** [comment]
I would like to implement this, but I have some questions to make sure I do the right thing.

The [paper](https://arxiv.org/abs/1206.5236) referenced in the description gives an exact synthesis for $2x2$ unitary matrices with entries in $\mathbb{Z}[\frac{1}{\sqrt{2}}, i]$ (and proves these are the only single-qubit matrices that can be exactly synthesized). For a practical algorithm, I would have to first approximate an arbitrary $2x2$ unitary with an unitary in $\mathbb{Z}[\frac{1}{\sqrt{2}}, i]$, and then apply the algorithm in the paper. The approximation part is non-trivial.

I have found two references that proceed along these lines: https://arxiv.org/abs/1212.6964 and https://arxiv.org/abs/1212.6253, and obtain a decomposition that is optimal or near-optimal in terms of T gates. Is it ok if I implement one of these? My preference is a bit towards https://arxiv.org/abs/1212.6253 because it has a simpler algorithm.

At the same time, better decompositions (with fewer T gates) can be obtained if one allows a repeat-until-success circuit with measurements, for example https://arxiv.org/abs/1404.5320. I saw this is one of the options offered by Qualtran.

In all decompositions the number of T-gates is of the form $K + c\log(\frac{1}{\epsilon})$, for constants K and $c$.

Let me know what you would prefer! @tanujkhattar 

@ACE07-Sev  please let me know your thoughts too!

**Comment by NoureldinYosri** [comment]
hi @codrut3, this is actually no longer needed since I implemented the current state of the art as part of the Qualtran library https://github.com/quantumlib/Qualtran/tree/main/qualtran/rotation_synthesis
